### PR TITLE
Arm backend: Document Ethos-U memory modes and add Ethos-U porting guide

### PR DESCRIPTION
**Pull request by gggekov**
Memory modes: The Shared_Sram, Sram_Only and Dedicated_Sram memory modes are specified in the compile spec and are tightly coupled with how the ethos-U scratch buffer and NN should be placed in the embedded application. Different memory modes profoundly impact the performance and memory footprint of the application and it is important to use the NPU in the most suitable memory mode for optimal performance.

Porting guide: A document explaining the key steps to port a new hardware target with an Ethos-U NPU to the Ethos-U backend in ExecuTorch

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

### CODE REVIEW

**Review by digantdesai — APPROVED** [review]
Didn't read too carefully, but love the diagrams :)
check the rendered docs on the PR before merging for formatting related issues mainly.

**Comment by zingo** [comment]
@GeorgeARM there seem to be a link problem see https://github.com/pytorch/executorch/actions/runs/17612710981/job/50038191324?pr=14144
This job need to be working for us to be able to merge.

Running command: docker exec -t 9a40a86ad68e5a1276c257b270a356fd5e3207fd65f96ad6cec20bfeabc4c560 /exec
++ '[' pull_request = pull_request ']'
++ echo 4d6209b35717896ed24f3c2228d11cbd6e601d64 4a85695c82f27748272d66df12d82a0f1458bed5
+ ./scripts/lint_xrefs.sh 4d6209b35717896ed24f3c2228d11cbd6e601d64 4a85695c82f27748272d66df12d82a0f1458bed5

docs/source/backends-arm-ethos-u.md:
FAIL examples/arm/ethos-u-porting-guide.md

examples/arm/ethos-u-porting-guide.md:
 OK  ../../docs/source/backends-arm-ethos-u.md
+ echo

+ echo 'Xref lint failed.'
Xref lint failed.
+ echo 'If this is a transient outage, you can bypass it by adding the `skip-xref-lint` label to your PR.'
If this is a transient outage, you can bypass it by adding the `skip-xref-lint` label to your PR.
+ echo 'Or add `@lint-ignore` somewhere on the same line as the reference you want to skip checking.'
Or add `@lint-ignore` somewhere on the same line as the reference you want to skip checking.

**Comment by zingo** [comment]
You seem to be able to reproduce and run on your maching if you run
./scripts/lint_xrefs.sh
on you patch (it will check all files)

**Comment by gggekov** [comment]
Thanks @digantdesai @zingo . The documentation renders correctly in https://docs-preview.pytorch.org/pytorch/executorch/14144/backends-arm-ethos-u.html 
with one exception. 

At the end of the md document, i am referring an absolute path that doesn't exist yet. 
```...the [Ethos-U porting guide](https://github.com/pytorch/executorch/blob/main/examples/arm/ethos-u-porting-guide.md).```

When the PR gets merged, the https://github.com/pytorch/executorch/blob/main/examples/arm/ethos-u-porting-guide.md webpage will be created and I assume the hyperlink will work correctly. Is this definitely true ? If yes, then I believe that is the reason for the failing CI test also.

File: docs/source/backends-arm-ethos-u.md [hunk 1]
```diff
@@ -72,17 +72,116 @@ with open("mv2_arm_ethos_u55.pte", "wb") as file:
...
 ```
 
+### Ethos-U memory modes
+The Ethos-U NPU provides two distinct memory interfaces:
+- One interface for **low-latency, high-bandwidth memory**
+Typically on-chip memory such as **SRAM**.
+- One interface for **higher-latency, lower-bandwidth memory**
+Typically external (off-chip) memory such as **Flash** or **DRAM**.
+
+On all Ethos-U NPUs(Ethos-U55, Ethos-U65, Ethos-U85), the low-latency interface is usually the SRAM of the SoC.
```
Comment (swolchok) [inline docs/source/backends-arm-ethos-u.md:82]: nit: missing space after "NPUs"

File: docs/source/backends-arm-ethos-u.md [hunk 2]
```diff
@@ -72,17 +72,116 @@ with open("mv2_arm_ethos_u55.pte", "wb") as file:
...
+The external memory type depends on the SoC:
+- On a low-power microcontorller, the external memory is usually Flash.
```
Comment (swolchok) [inline docs/source/backends-arm-ethos-u.md:84]: microcontroller

File: docs/source/backends-arm-ethos-u.md [hunk 3]
```diff
@@ -72,17 +72,116 @@ with open("mv2_arm_ethos_u55.pte", "wb") as file:
...
+- Neural Network resides in the External memory.
+- Ethos-U fast scratch buffer resides in the on-chip memory.
+- Characteristics:
+    - Used when the peak intermediate tensor is too big to fit into the on-chip memory.
+    - Enables silicon acceleration of large models.
+    - The NPU stores the results from the intermediate computations in the external memory.
+    - The dedicated SRAM acts as a software managed cache, improving performance by pre-fetching frequently accessed tensors to the on-chip memory.
+    - Available on Ethos-U65 and Ethos-U85.
+- Limitations:
+    - The SRAM space must be dedicated exculisely to the Ethos-U(the host processor should not access it).
```
Comment (swolchok) [inline docs/source/backends-arm-ethos-u.md:137]: nit: space after Ethos-U

File: docs/source/backends-arm-ethos-u.md [hunk 4]
```diff
@@ -72,17 +72,116 @@ with open("mv2_arm_ethos_u55.pte", "wb") as file:
...
+
+```
+Total SRAM used                               2467.27 KiB
+Total Off-chip Flash used                       12.20 KiB
+````
+The `Total SRAM used` indicates the peak SRAM utilization needed by the NPU in order to perform an inference. In the snippet above, the Ethos-U compiler requires 2467.27 KiB of SRAM in order to schedule the inference.
+Therefore, from an application standpoint, you need to ensure you have at least 2467.27 KiB of SRAM on the SoC to run this model. The Ethos-U compiler provides a scheduling algorithm allowing to
+lower the peak SRAM usage within reasonable limits, you need to add the `--optimise Size` or `--arena-cache-size` CLI options for to the compile spec. You can read more about the options of the
+Ethos-U compiler in the documentation [here](https://gitlab.arm.com/artificial-intelligence/ethos-u/ethos-u-vela/-/blob/main/OPTIONS.md#optimise). If the peak SRAM usage remains too high in
+Shared Sram memory mode, you would need to us the Dedicated Sram mode in order to store the Neural Network and the Ethos-U scratch buffer in the external memory.
```
Comment (swolchok) [inline docs/source/backends-arm-ethos-u.md:166]: all-caps SRAM?

File: docs/source/backends-arm-ethos-u.md [hunk 5]
```diff
@@ -72,17 +72,116 @@ with open("mv2_arm_ethos_u55.pte", "wb") as file:
...
+The main advantage of the Dedicated_Sram memory mode is that you can run large models and still benefit from the low-latency/high-bandwidth of the SRAM, used as a cache.
+
+It is important to highlight that when you specify a memory mode in the compile spec, in the runtime, the user is expected to place the scratch buffer and NN in the correct memory location.
+In other words, when you specify for ex. Shared Sram memory mode, the runtime application logic should place the ethos-U scratch buffer in the on-chip memory and the NN in the external memory for optimal performance.
```
Comment (swolchok) [inline docs/source/backends-arm-ethos-u.md:170]: SRAM, Ethos-U

File: examples/arm/ethos-u-porting-guide.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+any Ethos-U NPU.
+- Ethos-U fast scratch buffer -  a contiguous block of memory for the case when the Ethos-U scratch buffer and Neural Network are both in the external memory.
+Applicable only for Ethos-U65 and Ethos-U85 in Dedicated_Sram memory mode.
+
+### Ethos-U driver
+The key function of the Ethos-U driver enabling the interaction with the NPU is
+(ethosu_invoke_v3)[https://github.com/pytorch/executorch/blob/main/backends/arm/runtime/EthosUBackend.cpp#L324]. The `ethosu_invoke_v3` function takes as input a driver handle,
+a pointer towards  the command stream and the size of the command stream, base address as well as the size of the base addresses. For a system with Cortex-M, there is a 1:1
+mapping between base pointer and region, so we will pass three base pointers and each base pointer will correspond to one region. Then, as part of the compilation stage
+in `to_edge_transform_and_lower`, the Ethos-U compiler will generate command stream taking into account the three regions. In other words, at runtime, the ethos-U driver
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:39]: capitalize Ethos-U?

File: examples/arm/ethos-u-porting-guide.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+- Ethos-U scratch buffer - a contiguous block of memory used by the NPU to store the intermediate tensors produced and consumed during inference. Applicable for any Ethos-U NPU.
+- Neural Network - a contiguous block of memory holding constant data such as weights, biases, quantization parameters required to run an inference. Applicable for
+any Ethos-U NPU.
+- Ethos-U fast scratch buffer -  a contiguous block of memory for the case when the Ethos-U scratch buffer and Neural Network are both in the external memory.
+Applicable only for Ethos-U65 and Ethos-U85 in Dedicated_Sram memory mode.
+
+### Ethos-U driver
+The key function of the Ethos-U driver enabling the interaction with the NPU is
+(ethosu_invoke_v3)[https://github.com/pytorch/executorch/blob/main/backends/arm/runtime/EthosUBackend.cpp#L324]. The `ethosu_invoke_v3` function takes as input a driver handle,
+a pointer towards  the command stream and the size of the command stream, base address as well as the size of the base addresses. For a system with Cortex-M, there is a 1:1
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:37]: extra space after "torwards"

File: examples/arm/ethos-u-porting-guide.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+- The interrupt assignment for the Ethos-U. You also need to provide interrupt priority.
+
+In the `executorch/examples/arm/executor_runner/arm_executor_runner.cpp` sample application, we inherit the Corstone-300/Corstone-320 NPU initialization done in the [core-platform project](https://gitlab.arm.com/artificial-intelligence/ethos-u/ethos-u-core-platform).
+We include the core-platform project as a dependency in the `backends/arm/scripts/corstone_utils.cmake` script. Note that in [corstone_utils.sh](https://github.com/pytorch/executorch/blob/main/backends/arm/scripts/corstone_utils.cmake#L69)
+depending on whether we target Ethos-U55 or Ethos-U85, we include the  corresponding target from core-platform. Then, inside core-platform, the NPU base address and interrupt assignment are defined in the
+[target.cpp](https://gitlab.arm.com/artificial-intelligence/ethos-u/ethos-u-core-platform/-/blob/main/targets/corstone-320/target.cpp?ref_type=heads#L44) as per the memory map of the Corstone-300/Corstone-320.
+It's worth mentioning that the code in core-platform(code we reuse in the `examples/arm/executor_runner/arm_executor_runner.cpp`) also calls the `ethosu_init`
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:59]: space between core-platform and (

File: examples/arm/ethos-u-porting-guide.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+function to initialize the NPU. The `ethosu_init` function is [defined in the Ethos-U driver](https://gitlab.arm.com/artificial-intelligence/ethos-u/ethos-u-core-driver/-/blob/main/src/ethosu_driver.c?ref_type=heads#L409). The ethos-u driver itself is
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:60]: capitalize Ethos-U in "the ethos-u driver itself"?

File: examples/arm/ethos-u-porting-guide.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+included within the core-platform CMake. In other words, to initialize the NPU in the ExecuTorch executor runner application, we reuse the Ethos-U initialization that has been done in the core-platform project.
+Core-platform includes a [tutorial](https://gitlab.arm.com/artificial-intelligence/ethos-u/ethos-u-core-platform/-/blob/main/PORTING.md?ref_type=heads) on porting a new target. If you port your target to
+core-platform, you can then easily reuse it in the ExecuTorch runtime.
+
+Also, as explained in the comments in `backends/arm/scripts/corstone_utils.cmake`, note that REGIONCFG register of the Ethos-U controls the memory(on-chip or external memory) used by the NPU to
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:65]: missing space between memory and (

File: examples/arm/ethos-u-porting-guide.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+access the the Ethos-U scratch buffer, ML model and Ethos-U fast scratch buffer. The REGIONCFG is defined in the Ethos-U driver and you need to configure it differently depending on the memory mode.
+You can see in `backends/arm/scripts/corstone_utils.cmake` how we overwrite the register to the correct value depending on the memory mode.
+
+For the Corstone platforms, we make use of the Timing Adapters to model different memory latencies. The Timing Adapter is only applicable to the Corstone targets and should not be included for another SoC.
+
+## Corstone linker scripts
+
+The linker scripts point to the linker where to place various objects in memory when the application is loaded onto the target.
+In the `arm_executor_runner.cpp` application, we reuse the linker scripts from the core-platform project. Note that the Global Offset Table(.got symbols) needs to be 16-byte aligned. The linker scripts are highly specific to the memory map of system.
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:74]: missing space between Table and (
maybe s/the memory map of system/the memory map of the system/ ?

File: examples/arm/ethos-u-porting-guide.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+The CMSIS start-up code for Cortex-M is added as part of the build system of the core-platform applications.
+
+## Coupling between the AoT compile specification memory mode, linker script and the application logic
+It is important to note that when you specify a memory mode in the Python script to generate the pte file, in the runtime, the user is
+expected to place the scratch buffer and NN in the correct memory location.
+
+For example, if you generate a pte file with compile specification for Shared Sram, the scratch buffer should be placed in the SRAM and the NN in the external memory in the runtime application code. You can see we are following
+this approach in the `examples/arm/executor_runner/arm_executor_runner.cpp` example application. In the linker scripts for the application(`examples/arm/executor_runner/Corstone-320.ld` and
+`examples/arm/executor_runner/Corstone-300.ld`) we check the value of `ETHOSU_ARENA` to determine whether the ethos-u scratch buffer is placed in the on-chip memory or in the external memory. In this
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:87]: capitalize Ethos-U?

File: examples/arm/ethos-u-porting-guide.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+```
+unsigned char __attribute__((
+    section(".bss.tensor_arena"),
+    aligned(16))) temp_allocation_pool[temp_allocation_pool_size];
+```
+and the `.bss.tensor_arena` section is placed in the correct location in the memory map thanks to
+the `ETHOSU_ARENA` parameter.
+
+There is a tight coupling between the memory mode for the Ethos-U and the placement of the ethos-u scratch buffer,
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:101]: capitalize ethos-u?

File: examples/arm/ethos-u-porting-guide.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+SoC. The `arm_executor_runner.cpp` application built with the `examples/arm/run.sh` shell script and corresponding linker scripts are aimed to serve as
+example implementation for the correct placement of the various objects in memory.
+
+It's also worth mentioning that in the AoT Python flow, by default the input to the pte file is in FP32. Therefore, the pte file contains a Quantize node, an Ethos-U custom delegate and a Dequantize node.
+Sometimes, you may want to feed quantized input to the Ethos-U custom delegate straight away, for example if you have a camera input outputting RGB data in (u)int8. You can apply the `QuantizeInputs` and
+`QuantizeOutputs` passes in the AoT flow for that purpose. Here is a snippet showing how to achieve it:
+```
+edge_program_manager = to_edge_transform_and_lower(...)
+from executorch.exir.passes.quantize_io_pass import QuantizeInputs
+from executorch.exir.passes.quantize_io_pass import QuantizeOutputs
```
Comment (swolchok) [inline examples/arm/ethos-u-porting-guide.md:112]: just `from executorch.exir.passes.quantize_io_pass import QuantizeInputs, QuantizeOutputs`?
